### ISSUE TITLE
Comandlet «Get-Location» in parentheses

### ISSUE THREAD

**Issue by u20221022** [PowerShell/PowerShell#24023]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

I am experiencing a problem with adding of parentheses.

### Expected behavior

```console
`Get-Location -Stack | Get-Member`

   TypeName: System.Management.Automation.PathInfoStack

Name           MemberType Definition
----           ---------- ----------
Clear          Method     void Clear()
Contains       Method     bool Contains(System.Management.Automation.PathInfo item)
CopyTo         Method     void CopyTo(System.Management.Automation.PathInfo[] array, int arrayIndex), void ICollection.CopyTo(array array, int index)
EnsureCapacity Method     int EnsureCapacity(int capacity)
Equals         Method     bool Equals(System.Object obj)
GetEnumerator  Method     System.Collections.Generic.Stack`1+Enumerator[System.Management.Automation.PathInfo] GetEnumerator(), System.Collections.Generic.IEnum…
GetHashCode    Method     int GetHashCode()
GetType        Method     type GetType()
Peek           Method     System.Management.Automation.PathInfo Peek()
Pop            Method     System.Management.Automation.PathInfo Pop()
Push           Method     void Push(System.Management.Automation.PathInfo item)
ToArray        Method     System.Management.Automation.PathInfo[] ToArray()
ToString       Method     string ToString()
TrimExcess     Method     void TrimExcess()
TryPeek        Method     bool TryPeek([ref] System.Management.Automation.PathInfo result)
TryPop         Method     bool TryPop([ref] System.Management.Automation.PathInfo result)
Count          Property   int Count {get;}
IsSynchronized Property   bool IsSynchronized {get;}
Name           Property   string Name {get;}
SyncRoot       Property   System.Object SyncRoot {get;}
```

### Actual behavior

```console
`(Get-Location -Stack) | Get-Member`
Get-Member: You must specify an object for the Get-Member cmdlet.
```

### Error details

```console
`Get-Error`

Exception             :
    Type    : System.InvalidOperationException
    Message : You must specify an object for the Get-Member cmdlet.
    HResult : -2146233079
CategoryInfo          : CloseError: (:) [Get-Member], InvalidOperationException
FullyQualifiedErrorId : NoObjectInGetMember,Microsoft.PowerShell.Commands.GetMemberCommand
InvocationInfo        :
    MyCommand        : Get-Member
    ScriptLineNumber : 1
    OffsetInLine     : 25
    HistoryId        : 144
    Line             : (Get-Location -Stack) | Get-Member
    Statement        : Get-Member
    PositionMessage  : At line:1 char:25
                       + (Get-Location -Stack) | Get-Member
                       +                         ~~~~~~~~~~
    InvocationName   : Get-Member
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
```

### Environment data

```powershell
`$PSVersionTable`

Name                           Value
----                           -----
PSVersion                      7.4.3
PSEdition                      Core
GitCommitId                    7.4.3
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by mklement0** [comment]
`Get-Location -Stack` is unusual in that it emits an enumerable (of type [`System.Management.Automation.PathInfoStack`](https://learn.microsoft.com/en-US/dotnet/api/System.Management.Automation.PathInfoStack)) _as a whole_ instead of _auto-enumerating_ its elements.

Enclosing any command in `(...)` _forces_ auto-enumeration.

If the enumerable has _no elements_ - as in your case - `Get-Member` by design complains, as then receives _no input_.  
It is the equivalent of: `@() | Get-Member`

**Comment by u20221022** [comment]
Thank you, but…

> is unusual

If it is not a mistake, how could it have been guessed?  How can method `.GetType()` be applied if parentheses are not available?
`Get-Location -Stack -Is [Management.Automation.PathInfoStack[]]`
Get-Location: A parameter cannot be found that matches parameter name 'Is'.

`[Management.Automation.PathInfoStack[]] $pisPIS = Get-Location -Stack`
`$pisPIS.GetType()`
```
IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     PathInfoStack[]                          System.Array
```

**Comment by u20221022** [comment]
> Enclosing any command

Any command or any value?  What does that mean:
`(Get-Location -Stack).GetEnumerator | Get-Member`

```
   TypeName: System.Management.Automation.PSMethod`1[[System.Management.Automation.MethodGroup`2[[System.Func`1[[System.Collections.Generic.Stack`1+Enumerator[[S
ystem.Management.Automation.PathInfo, System.Management.Automation, Version=7.4.3.500, Culture=neutral, PublicKeyToken=31bf3856ad364e35]], System.Collections, Ve
rsion=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]], System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Sy
stem.Management.Automation.MethodGroup`2[[System.Func`1[[System.Collections.Generic.IEnumerator`1[[System.Management.Automation.PathInfo, System.Management.Autom
ation, Version=7.4.3.500, Culture=neutral, PublicKeyToken=31bf3856ad364e35]], System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7be
a7798e]], System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Func`1[[System.Collections.IEnumerator, System.Priva
te.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]], System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d
7bea7798e]], System.Management.Automation, Version=7.4.3.500, Culture=neutral, PublicKeyToken=31bf3856ad364e35]], System.Management.Automation, Version=7.4.3.500
, Culture=neutral, PublicKeyToken=31bf3856ad364e35]]

Name                MemberType Definition
----                ---------- ----------
Copy                Method     System.Management.Automation.PSMemberInfo Copy()
Equals              Method     bool Equals(System.Object obj)
GetHashCode         Method     int GetHashCode()
GetType             Method     type GetType()
Invoke              Method     System.Object Invoke(Params System.Object[] arguments)
ToString            Method     string ToString()
IsInstance          Property   bool IsInstance {get;}
MemberType          Property   System.Management.Automation.PSMemberTypes MemberType {get;}
Name                Property   string Name {get;}
OverloadDefinitions Property   System.Collections.ObjectModel.Collection[string] OverloadDefinitions {get;}
TypeNameOfValue     Property   string TypeNameOfValue {get;}
Value               Property   System.Object Value {get;set;}
```

**Comment by rhubarb-geek-nz** [comment]
> (Get-Location -Stack).GetEnumerator | Get-Member

Do you want to call GetEnumerator? If so you need brackets, eg

```
((Get-Location -Stack).GetEnumerator()).Path
```

**Comment by u20221022** [comment]
> need brackets

Sorry, but it do not need, it unacceptable (at my computer):

`(Get-Location -Stack).GetEnumerator() | Get-Member`
Get-Member: You must specify an object for the Get-Member cmdlet.

`((Get-Location -Stack).GetEnumerator()).Path`
Nothing.

`((Get-Location -Stack).GetEnumerator()).Path.GetType()`
InvalidOperation: You cannot call a method on a null-valued expression.

**Comment by u20221022** [comment]
@mklement0
Sorry, can you tell me how to check if the stack is empty?  Comparison with the `$Null` does not work.
Variant №1 – method «.Count»: `(Get-Location -Stack).Count -EQ 0`.
Variant №2 – ???
Thank you!

**Comment by rhubarb-geek-nz** [comment]
> can you tell me how to check if the stack is empty?

```
 ((Get-Location -Stack) | Measure-Object).Count
```

A bit cumbersome but appears to work

**Comment by u20221022** [comment]
@rhubarb-geek-nz
That is exactly what I have done so far.  But I found this method unattractive.

**Comment by rhubarb-geek-nz** [comment]
> @rhubarb-geek-nz That is exactly what I have done so far. But I found this method unattractive.

How about

```
if (-not (Get-Location -Stack).Count)
{
    'empty'
}
```

**Comment by u20221022** [comment]
@mklement0
Sorry, if you know.  What is happening withOUT of parentheses?  My imagination is lacking.

`$objGLS = Get-Location -Stack`

`$objGLS.Count`
0

`$objGLS[0]`
Nothing.

`$objGLS[1]`
Nothing.

`$Null -EQ $objGLS[0]`
False

`$Null -EQ $objGLS[1]`
True

`$objGLS[0].GetType()`

```
IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     False    PathInfoStack                            System.Collections.Generic.Stack`1[System.Management.Automation.PathInfo]
```

**Comment by mklement0** [comment]
The following works (returns `$true`):

`(Get-Location -Stack) -Is [Management.Automation.PathInfoStack]`

In the context of an _expression_, `(...)` does _not_ perform auto-enumeration.
By contrast, in the _pipeline_ (e.g. `(Get-Location -Stack) | Get-Member`, it does.

> can you tell me how to check if the stack is empty? 

Yes, `0 -eq (Get-Location -Stack).Count` works.

There's no need to use `.GetEnumerator()`.

**Comment by u20221022** [comment]
> In the context of an _expression_, `(...)` does _not_ perform auto-enumeration.
> By contrast, in the _pipeline_ (e.g. `(Get-Location -Stack) | Get-Member`, it does.

Thank you, it is important.

**Comment by rhubarb-geek-nz** [comment]
Can you explain what you are trying to achieve? I can't think of any reason why one would need to know the depth of the stack. The simple rule should be that a Push-Location should match a Pop-Location, and you should only Pop-Location if you did the Push-Location.

I use the simple pattern of

```
Push-Location 'foo'

try
{
    ... do stuff....
}
finally
{
    Pop-Location
}
```

So the Pop-Location always matches the Push-Location even in the case of an error or thrown exception.

**Comment by u20221022** [comment]
> I use the simple pattern

But I use more simple and more reliable way and do not care about what happens (what errors are present) in the subroutines (this is real citation from my program):
```
Try/Catch …
} Finally {	# 
    While ((Get-Location -Stack).Count -GT 0) {	# 
        Pop-Location
    }	# End_While 
    Write-…
}	# End_Try 
```

By the way, a simple example showing that you wrote a silly thing and have unpaired «Pop»:
`Push-Location 'foo'`
Push-Location: Cannot find path '…\foo' because it does not exist.

And I do not breed short lines (one word per line, Dumas-father style, who was paid line by line) because my neck is not government property.  Respect those who will read your writing.

**Comment by rhubarb-geek-nz** [comment]
Your finally has a problem if I call Push-Location before calling your program, your program would Pop what I pushed.

In my scripts I always uses $ErrorActionPreference = Stop so if Push-Location failed it would have raised an error and not continued to the try/finally.

**Comment by 237dmitry** [comment]
This works then Stack greater 0 but returns PathInfo instead of PathInfoStack:

```
 $ (Get-Location -stack) | gm
Get-Member: You must specify an object for the Get-Member cmdlet.
 $ pushd
 $ (Get-Location -stack) | gm

   TypeName: System.Management.Automation.PathInfo

Name         MemberType Definition
----         ---------- ----------
Equals       Method     bool Equals(System.Object obj)
GetHashCode  Method     int GetHashCode()
GetType      Method     type GetType()
ToString     Method     string ToString()
Drive        Property   System.Management.Automation.PSDriveInfo Drive {get;}
Path         Property   string Path {get;}
Provider     Property   System.Management.Automation.ProviderInfo Provider {get;}
ProviderPath Property   string ProviderPath {get;}
```

**Comment by u20221022** [comment]
> This works then Stack greater 0

Count of Stack?  When (Get-Location -Stack)>0?  That is what we have already discussed.

**Comment by u20221022** [comment]
> This works then Stack greater 0

Count of Stack?  When (Get-Location -Stack).Count>0?  That is what we have already discussed.

**Comment by 237dmitry** [comment]
> That is what we have already discussed

Where? Why this returns PathInfo instead of PathInfoStack?

**Comment by u20221022** [comment]
> > That is what we have already discussed
> 
> Where? Why this returns PathInfo instead of PathInfoStack?

In first comment (https://github.com/PowerShell/PowerShell/issues/24023#issuecomment-2215504774).

**Comment by 237dmitry** [comment]
> If you use the Stack or StackName parameters, this cmdlet returns a StackInfo object. Otherwise, it returns a PathInfo object.

Look:
![ArcoLinux_2024-07-09_11-46-39](https://github.com/PowerShell/PowerShell/assets/78153320/c8a8fc8c-c32e-4fbb-8699-cfcf9d39e186)

**Comment by u20221022** [comment]
> Look:

Who wrote the text of help?  Did I?
Why in first line – «PathInfo» & «PathInfoStack», but in second line – «StackInfo» & «PathInfo»?  «Info Stack» ≡ «Stack Info»?

**Comment by 237dmitry** [comment]
I just showed that when the push stack is greater than zero, Get-Member receives a PathInfo object. And this is in the context of your issue, with which I agree. I don't understand why parentheses affect the output of PathInfoStack, but have no effect on the output of PathInfo.

>  And I not remember aliases, sorry.  «gu» – what is this?

```
 $ Get-Alias gu

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Alias           gu -> Get-Unique                                              

```

**Comment by u20221022** [comment]
@237dmitry
PathInfo – one ELEMENT of stack.
PathInfoStack – WHOLE stack as whole.

**Comment by 237dmitry** [comment]
> PathInfo – one ELEMENT of stack

Could you explain what is an  element of stack? `PathInfoStack` does not contain `PathInfo` property:

```powershell
$ (Get-Location -Stack | Get-Member -MemberType Property).Name
Capacity
Count
IsSynchronized
Name
SyncRoot
```

**Comment by u20221022** [comment]
@237dmitry
Try it this way (line breaks in the console).  First – withOUT of parentheses, second – with parentheses.
```
>> Get-Location -Stack | 
>>     ForEach-Object {
>>         If (    $_ -Is [System.Management.Automation.PathInfoStack]) {
>>             Write-Host 1
>>         } ElseIf ($_ -Is [System.Management.Automation.PathInfo]) {
>>             Write-Host 2
>>         } # End_If
>>     } # End_ForEach-Object
```
1

```
>> (Get-Location -Stack) | 
>>     ForEach-Object {
>>         If (    $_ -Is [System.Management.Automation.PathInfoStack]) {
>>             Write-Host 1
>>         } ElseIf ($_ -Is [System.Management.Automation.PathInfo]) {
>>             Write-Host 2
>>         } # End_If
>>     } # End_ForEach-Object
```
2

**Comment by u20221022** [comment]
> Could you explain

Not, not, not.  I know Fortran, Algol, Cobol etc, but NOT PowerShell.
Can you explain what the method «Clear» does?  The FIRST method in list.  What exactly is it destroying?  I do not see any result.

**Comment by jhoneill** [comment]
@u20221022 
This might make it clearer 
![image](https://github.com/PowerShell/PowerShell/assets/16209593/919b6791-7334-47d3-af47-692ef2fcf60c)

`Get-Location -stack`  returns the stack object: always one stack 

`(Get-Location -stack)`  returns the items on stack. 

It is the same if we assign to a variable. 

![image](https://github.com/PowerShell/PowerShell/assets/16209593/d34a9401-5d64-4292-b589-9f99d9b2d102)

With most objects 
`$x`and `($x)` give the same thing.  But `PathInfoStack` is a special case and **the special cases are not easily discoverable**.

**Comment by u20221022** [comment]
> the special cases are not easily discoverable

That is for sure!  Difficult tasks are always difficult to solve.

**Comment by rhubarb-geek-nz** [comment]
> But I use more simple .. and do not care about what happens ... (this is real citation from my program):

PowerShell is not like bash or other shells. The script executes in the same process as the caller, the location stack does not belong to your script but to the current thread. So on entry, if the called had called Push-Location before calling your script the depth the the stack would be non-zero. So to use your pattern you need to capture the depth of the stack when your script starts and then in the finally pop until the same depth is reached.

 ```
$OriginalStackCount = (Get-Location -Stack).Count

try
{
    foo ...
    bar ...
}
finally
{ 
    While ((Get-Location -Stack).Count -GT $OriginalStackCount) 
    { 
         Pop-Location
    }
}
```

If you don't take this measure the caller will not have the expected location stack on your scripts return/exit.

**Comment by rhubarb-geek-nz** [comment]
> Can you explain what the method «Clear» does?

It empties the contents of the stack

[Stack.Clear](https://learn.microsoft.com/en-us/dotnet/api/system.collections.stack.clear)

**Comment by rhubarb-geek-nz** [comment]
> With most objects `$x`and `($x)` give the same thing. But `PathInfoStack` is a special case and **the special cases are not easily discoverable**.

Depends how curious you are.

```
Title         Invoke        ForEach       BracketsForEach Assign        Index         IList IEnumerable
-----         ------        -------       --------------- ------        -----         ----- -----------
string        String        String        String          String        Char          False        True
Stack         Stack         Stack         Int32           Stack         Stack         False        True
PathInfoStack PathInfoStack PathInfoStack PathInfo        PathInfoStack PathInfoStack False        True
ArrayList     ArrayList     ArrayList     Int32           ArrayList     Int32          True        True
array         Object[]      Object[]      Int32           Object[]      Int32          True        True
int[]         Int32[]       Int32[]       Int32           Int32[]       Int32          True        True
char[]        Char[]        Char[]        Char            Char[]        Char           True        True
JSON array    Object[]      Object[]      Int64           Object[]      Int64          True        True
```

There is nothing special about PathInfoStack other that it is not a form of array. It is IEnumerable just like string. Unlike String, PowerShell does not enumerate strings to give you char, but does enumerate PathInfoStack to give you PathInfo.

One of the original sins of PowerShell is trying to pretend that a list of one item is actually one item, and then trying to pretend that one item is actually a list containing one element. This is in contrast to SQL API where a result set is always a list of records, even if there is only one record. Where this catches PowerShell out is when the one item that a cmdlet returns is an IEnumerable, or the single object you are outputting is actually an IEnumerable and when you write it you end up with multiple output records.

**Comment by jhoneill** [comment]
> But I use more simple and more reliable way and do not care about what happens (what errors are present) in the subroutines (this is real citation from my program):
> 
> ```
> Try/Catch …
> } Finally {	# 
>     While ((Get-Location -Stack).Count -GT 0) {	# 
>         Pop-Location
>     }	# End_While 
>     Write-…
> }	# End_Try 
> ```


This is a bad thing to do. Your script does not own the stack, and emptying it is rude, like clearing the terminal at the start of a script.  My profile replaces `CD` with something that has its own tab expansion, and which calls `Push-Location`   so for **me** `cd = [tab]` cycles through the stack, so I can move to somewhere on the stack without popping entries -giving the effect of `cd -` and `cd +` , so your script would put me back in the directory where PowerShell started.   

IF you script does a directory traversal with `Push-Location`, you can  use  `push-Location someDir  -StackName MyStack` but if if `mystack` is empty `get-Location -StackName Mystack` will give an error, so popping every location doesn't work well.  
You're better putting `$initialDir = $pwd` at the start of the script and `Set-Location $initialDir` in the `finally` block. 



> > the special cases are not easily discoverable. 
> Depends how curious you are.

Or your definition of "easily" :-)  If I take an array, an array list, something like the keycollection member of a hash table and  pipe into something I get the result of enumerating the collection not the collection object. It's not as a simple of if is iEnumerable enumerate it (unless it's a string).  

```
> $h = @{1="First";2="Second"}
> $h.Keys | gm
   TypeName: System.Int32
   # Snip

> gm -InputObject $h.keys
   TypeName: System.Collections.Hashtable+KeyCollection
   # Snip
```

**Comment by rhubarb-geek-nz** [comment]
> It's not as a simple of if is iEnumerable enumerate it (unless it's a string).

Fair enough, however ICollection.GetEnumerable() is inherited from IEnumeable.GetEnumable()

**Comment by mklement0** [comment]
Auto-enumeration _is_ about `IEnumerable`, but with selective exceptions, and one addition:

Auto-enumeration applies to types that implement the [`IEnumerable`](https://learn.microsoft.com/en-US/dotnet/api/System.Collections.IEnumerable) interface or its generic counterpart, but there are a few _exceptions_: 
* strings
* dictionaries (types implementing `IDictionary` or its generic counterpart, including hashtables)
* `System.Xml.XmlNode` instances. 

Additionally, `System.Data.DataTable` is enumerated too, despite not implementing `IEnumerable` itself, via its `.Rows` property:

https://github.com/PowerShell/PowerShell/blob/dcf67544451425c98580d79a8381e9a07775c86d/src/System.Management.Automation/engine/LanguagePrimitives.cs#L528-L544

**Comment by u20221022** [comment]
@jhoneill
> your script would put me back in the directory where PowerShell started

MY script put YOU?  What do YOU have to do with MY script?  You do not need to stick your fingers in an electrical sockets and you do not need to call from other programs what is NOT designed for such calls.  NOT designed.  Can you understand this?
I gave an example of the LAST lines of my script.  LAST lines.  After executing them, I manually run the script again.  And I want the re-execution to take place in the same environment as the initial execution.
If you do not understand the difference between a head (or main) module and a library of modules intended for external calls, get away from me.

**Comment by rhubarb-geek-nz** [comment]
> I gave an example of the LAST lines of my script.

Indeed, and that is where the problem lies.

```
    While ((Get-Location -Stack).Count -GT 0) {	# 
        Pop-Location
    }	# End_While 
```

You code says `-gt 0`, so is removing _all_ items from the location stack.

> If you do not understand the difference between a head (or main) module 

There isn't really such a thing in PowerShell because all scripts run in the same process, so one script calling another run in the same process and share the same thread and location stack.

**Comment by u20221022** [comment]
@jhoneill
> Or your definition of "easily"

Where did you see my definition?  You made it up and you are trying to attribute it to me?

**Comment by jhoneill** [comment]
> @jhoneill
> 
> > Or your definition of "easily"
> 
> Where did you see my definition? You made it up and you are trying to attribute it to me?

No.  Please try to read what is said, tone down the aggression and if need be read the code of conduct (link at the the bottom).

I said 
>  PathInfoStack is a special case and the special cases are not easily discoverable.

@rhubarb-geek-nz   replied
>>  not easily discoverable

> Depends how curious you are.

and I said (well if you say it is **easily** discoverable depending on your curiosity then the other variable is)

>  your definition of "easily"

**Comment by u20221022** [comment]
@jhoneill
Sorry my translator.  I do not speak English and I can not see any aggression.  Nor do I see any I do not see any guilt towards those who my ancestors could not have oppressed because they lived on another continent.
Please, indicate, where you see MY definition.

**Comment by jhoneill** [comment]
> @jhoneill
> 
> > your script would put me back in the directory where PowerShell started
> 
> MY script put YOU?  What do YOU have to do with MY script?  You do not need to stick your fingers in an electrical sockets 
Again,  try to stay respectful.

> you do not need to call from other programs what is NOT designed for such calls.  
Your program can't be called from the > of someone who uses Push-Location in their shell. 

> NOT designed. 
Clear it wasn't  

> Can you understand this? 

Better than you seem to. 

> I gave an example of the LAST lines of my script.  LAST lines. 

Those lines are in a finally block. They will always run and they will always trash the stack  

Let's say you do this. 
```
C:\ > Push-Location   dir_where_your_script_is 
dir_where_your_script_is> .\your_script.ps1
C:\>  [up arrow]    .\your_script.ps1
not found error message
```


>  After executing them, I manually run the script again.  And I want the re-execution to take place in the same environment as the initial execution.

But environment is **only the same if you can absolutely guarantee that the stack was empty when your script ran**. 

> If you do not understand the difference between a head (or main) module and a library of modules intended for external calls, get away from me.

And until you understand the assumptions you are making about the environment you're running in, never, ever let another soul run you code. Then you and delete the back scroll buffer in the  terminal, mess with the command history, play with the directory stack.

**Comment by jhoneill** [comment]
> @jhoneill Sorry my translator.  I do not speak English and I can not see any aggression.  Nor do I see any I do not see any guilt towards those who my ancestors could not have oppressed because they lived on another continent. Please, indicate, where you see MY definition.

What on earth you are talking about with the ancestor guilt ? 

@rhubarb-geek-nz  responded to what I said about "easy" , I replied back. 
Nothing in that exchange had any reference to what **you** thought was or was not easy.  
The humour might have been lost in translation.  He said, in effect, "If you are curious, then discovering that seems easy", and I said "we must mean different things by 'easy' then "

**Comment by rhubarb-geek-nz** [comment]
> I know Fortran, Algol, Cobol etc, but NOT PowerShell.

So it is worth listening to those who do know about the execution environment of PowerShell, it is nothing like Fortran, Algol, Cobol or even bash or batch files. There is no such thing as a main module, the only main there is in the pwsh.exe itself. Any script can be called by any other script, there is no hierarchy other than actual modules, but even they can contain scripts.

The execution environment, the process, the threads, the location stack are all shared and scripts have to be cooperative to work well together. Anyone can call Push-Location before calling your script, and if they do so they have a right to expect that Pop-Location will take them back to where they were, which won't happen if you pop more entries from the stack than you pushed.

**Comment by u20221022** [comment]
@jhoneill
Sorry, I can not be sure I understand the jumble of your quotes correctly.  If you want to brag, write to my mail address in my native language.  MY NATIVE.

>if you can absolutely guarantee that the stack was empty when your script ran
What is so hard about it if the computer belongs to me and is completely at my mercy?  If you are so smart, try calling at least one of those scripts that are as files on my computer.  After that I am ready to rewrite them all according to your recommendations.  In the meantime, I write the way I need, not you need.

**Comment by u20221022** [comment]
@jhoneill
> What on earth you are talking about with the ancestor guilt ?

About BLM.

**Comment by u20221022** [comment]
@jhoneill
> responded to what I said about "easy" , I replied back.

Sorry, you are start from citation from me and I make mistake.  Sorry.

**Comment by u20221022** [comment]
@jhoneill
@mklement0

Some time ago someone rebuked me for using the «Private» visibility limiter.  He even stated that he had never used such a limiter in his life.
Now another «alternatively gifted» claims that all scripts are crammed into one process (what a surprise! who would have thought it?)
If I were paying attention to them, I would be flailing from side to side like a drunk driver.
P.S.  Moving through the stack allows you to see where the program is during step-by-step execution.  After the rough debugging is complete, it is quite possible to exclude such methods completely and specify full paths.  It will be even faster, as far as I understand.

**Comment by rhubarb-geek-nz** [comment]
> Update: P.P.S.  «Alternatively gifted» people either have never heard of named stacks themselves, or they think that I will never master such complex wisdom.

If you were using named stacks then you would see there is no need to pop off all the locations. All you would need to do is create a new named stack and set it to be your default stack, then in your finally restore the original stack by name and set original location.

**Comment by jhoneill** [comment]
> @jhoneill @mklement0
> 
> Some time ago someone rebuked me for using the «Private» visibility limiter.  He even stated that he had never used such a limiter in his life. Now another «alternatively gifted» claims that all scripts are crammed into one process (what a surprise! who would have thought it?) If I were paying attention to them, I would be flailing from side to side like a drunk driver. P.S.  Moving through the stack allows you to see where the program is during step-by-step execution.  After the rough debugging is complete, it is quite possible to exclude such methods completely and specify full paths.  It will be even faster, as far as I understand.
> 
> Update: P.P.S.  «Alternatively gifted» people either have never heard of named stacks themselves, or they think that I will never master such complex wisdom.  What prevents those unfortunate scripts that have to cram together like beggars in one single process from distributing stacks with different names? Or those who have no Fortran programming experience can not even understand that? By the way, the phrase about Fortran was a joke and was used in a purely personal conversation.  Only an extremely dishonest creature could pick on it.

At some point you really should read the code of conduct 

You have been given advice about best practice in PowerShell,  if you chose to say your scripts run only in a very tightly controlled environment where you can use worst practice that's your call. But these threads form a kind of knowledge base so people with better understanding than yours **will** correct what you have said. 
You know Algol and and Fortran ? Good for you but don't expect anyone to be impressed. I don't have the widest knowledge here and I've programmed in Fortran, Basic, Pascal, Forth, Occam, Snobol, Lisp, Assembler, C#,  as well as macro languages, scripting languages, and have even considered starting a Python victims support group :-)  

The use of `private` is poorly understood, it doesn't stop `Get-Variable` accessing the variable from a child scope, and its use can usually be avoided. I've seen code where everything is marked `global` or `script` or `private` to help the programmer know what was used outside its own scope but it's unnecessary.   

Named  stacks would be a solution for your bad practice, EXCEPT  the stack is discarded when empty so simply modifying your code to  
```
while ((get-location -stackname MyStack).count -gt 0) {pop-location -stackname mystack}
``` 

Will cause an error (incidentally an experienced PowerShell users know there is no need to test for Zero / Non-zero since a non zero count will be treated as `true`,  it's one of those signs like using a lot of `write-host` commands or ToString() when not needed  that someone is thinking in BASIC or ForTran but writing the actual code in PowerShell )
You could do something like `do {Pop-Location -StackName MyStack -ErrorAction SilentlyContinue} while ($?)`

**Comment by SteveL-MSFT** [comment]
@u20221022 as noted, stay within the [Code of Conduct](https://github.com/PowerShell/PowerShell/blob/master/CODE_OF_CONDUCT.md) so we can have productive discussion keeping in mind that others are trying to help you.  You may be attempting some humor, but it doesn't translate well in written form and also remember that not everyone that participates here have English as their native language.

**Comment by iRon7** [comment]
TL;DR, you might simply use the [comma operator `,`](https://learn.microsoft.com/powershell/module/microsoft.powershell.core/about/about_operators#comma-operator-)

```PowerShell
,(Get-Location -Stack) | Get-Member

   TypeName: System.Management.Automation.PathInfoStack

Name           MemberType Definition
----           ---------- ----------
Clear          Method     void Clear()
Contains       Method     bool Contains(System.Management.Automation.PathInfo item)
CopyTo         Method     void CopyTo(System.Management.Automation.PathInfo[] array, int arrayIndex), void ICollection.CopyTo(array array, i…
EnsureCapacity Method     int EnsureCapacity(int capacity)
Equals         Method     bool Equals(System.Object obj)
GetEnumerator  Method     System.Collections.Generic.Stack`1+Enumerator[System.Management.Automation.PathInfo] GetEnumerator(), System.Colle…
GetHashCode    Method     int GetHashCode()
GetType        Method     type GetType()
Peek           Method     System.Management.Automation.PathInfo Peek()
Pop            Method     System.Management.Automation.PathInfo Pop()
Push           Method     void Push(System.Management.Automation.PathInfo item)
ToArray        Method     System.Management.Automation.PathInfo[] ToArray()
ToString       Method     string ToString()
TrimExcess     Method     void TrimExcess()
TryPeek        Method     bool TryPeek([ref] System.Management.Automation.PathInfo result)
TryPop         Method     bool TryPop([ref] System.Management.Automation.PathInfo result)
Count          Property   int Count {get;}
IsSynchronized Property   bool IsSynchronized {get;}
Name           Property   string Name {get;}
SyncRoot       Property   System.Object SyncRoot {get;}
```

```PowerShell
,(Get-Location -Stack) |
    ForEach-Object {
        If ($_ -Is [System.Management.Automation.PathInfoStack]) {
            Write-Host 1
        } ElseIf ($_ -Is [System.Management.Automation.PathInfo]) {
            Write-Host 2
        } # End_If
    } # End_ForEach-Object

1
```
